File: app/src/main/java/io/github/lamtran/moviebooking/SeatAdapter.java
package io.github.lamtran.moviebooking;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.github.lamtran.moviebooking.model.state.AvailableState;
import io.github.lamtran.moviebooking.model.state.EmptyState;
import io.github.lamtran.moviebooking.model.state.ReservedState;
import io.github.lamtran.moviebooking.model.state.SelectedState;
import io.github.lamtran.moviebooking.model.state.State;
import io.github.lamtran.moviebooking.ui.viewholder.SeatViewHolder;
import io.github.lamtran.moviebooking.ui.viewholder.SeatViewHolderFactory;

/**
 * Created by lam on 2/3/17.
 */

public class SeatAdapter extends RecyclerView.Adapter<SeatViewHolder> {

  private final Map<Integer, SeatViewHolderFactory> mSeatViewHolderFactoryMap;

  private List<Seat> mSeats;

  public SeatAdapter(Map<Integer, SeatViewHolderFactory> seatViewHolderFactoryMap) {
    mSeats = new ArrayList<>();
    mSeatViewHolderFactoryMap = seatViewHolderFactoryMap;
  }

  @Override
  public int getItemCount() {
    return mSeats.size();
  }

  @Override
  public int getItemViewType(int position) {
    Seat seat = mSeats.get(position);
    State state = seat.getState();
    if (state instanceof EmptyState) {
      return Seat.TYPE_EMPTY;
    } else if (state instanceof ReservedState) {
      return Seat.TYPE_RESERVED;
    } else if (state instanceof AvailableState) {
      return Seat.TYPE_AVAILABLE;
    } else if (state instanceof SelectedState) {
      return Seat.TYPE_SELECTED;
    } else {
      return Seat.TYPE_UNKNOWN;
    }
  }

  @Override
  public void onBindViewHolder(SeatViewHolder holder, int position) {
    holder.bind(mSeats.get(position));
  }

  @Override
  public SeatViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    return mSeatViewHolderFactoryMap.get(viewType).createViewHolder(parent);
  }

  public void setSeats(List<Seat> seats) {
    mSeats = seats;
    notifyDataSetChanged();
  }
}

File: app/src/main/java/io/github/lamtran/moviebooking2/SeatAdapter.kt
package io.github.lamtran.moviebooking2

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

class SeatAdapter<VH : SeatAdapter.SeatViewHolder> : RecyclerView.Adapter<VH>() {

    private val seats: List<Seat> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemCount(): Int = seats.size

    override fun onBindViewHolder(holder: VH, position: Int) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    open class SeatViewHolder(open val view: View) : RecyclerView.ViewHolder(view) {

    }

    class EmptySeatViewHolder(override val view: View) : SeatViewHolder(view) {

    }

    class ReservedSeatViewHolder(override val view: View) : SeatViewHolder(view) {

    }

    class ChosenSeatViewHolder(override val view: View) : SeatViewHolder(view) {

    }

    class AvailableSeatViewHolder(override val view: View) : SeatViewHolder(view) {

    }
}
